test_force_release regression on Questa with FLI

Introduced #1403.

```
 $ make SIM=questa COMPILE_ARGS="+acc" SIM_ARGS="-t ps" TOPLEVEL_LANG=vhdl
make results.xml
make[1]: Entering directory `.../cocotb/tests/test_cases/test_force_release'
mkdir -p sim_build
set -o pipefail; LD_LIBRARY_PATH=.../lib/python3.7/site-packages/cocotb/libs/modelsim: MODULE=test_force_release TESTCASE= TOPLEVEL="work.sample_module" COCOTB_SIM=1 \
GPI_EXTRA= TOPLEVEL_LANG=vhdl PYTHONPATH=.../lib/python3.7/site-packages/cocotb/libs/modelsim:.../cocotb/tests/test_cases/test_force_release:.../lib/python3.7/site-packages:... \
.../bin/vsim -c -64  -do sim_build/runsim.do  2>&1 | tee sim_build/sim.log
Reading pref.tcl

# 2019.1

# do sim_build/runsim.do
# QuestaSim-64 vmap 2019.1 Lib Mapping Utility 2019.01 Jan  1 2019
# vmap -c 
# Copying .../linux_x86_64/../modelsim.ini to modelsim.ini
# QuestaSim-64 vmap 2019.1 Lib Mapping Utility 2019.01 Jan  1 2019
# vmap work sim_build/work 
# Modifying modelsim.ini
# QuestaSim-64 vcom 2019.1 Compiler 2019.01 Jan  1 2019
# Start time: 16:46:50 on Mar 03,2020
# vcom -work work "+acc" .../cocotb/tests/test_cases/test_force_release/../../../tests/designs/sample_module/sample_module_pack.vhdl .../cocotb/tests/test_cases/test_force_release/../../../tests/designs/sample_module/sample_module_1.vhdl .../cocotb/tests/test_cases/test_force_release/../../../tests/designs/sample_module/sample_module.vhdl 
# -- Loading package STANDARD
# -- Loading package TEXTIO
# -- Loading package std_logic_1164
# -- Loading package NUMERIC_STD
# -- Compiling package sample_module_pack
# -- Compiling package body sample_module_pack
# -- Loading package sample_module_pack
# -- Compiling entity sample_module_1
# -- Compiling architecture impl of sample_module_1
# -- Loading package sample_module_pack
# -- Compiling entity sample_module
# -- Compiling architecture impl of sample_module
# End time: 16:46:50 on Mar 03,2020, Elapsed time: 0:00:00
# Errors: 0, Warnings: 0
# vsim -t ps -onfinish exit -foreign "cocotb_init .../lib/python3.7/site-packages/cocotb/libs/modelsim/libfli.so" sim_build/work.sample_module 
# Start time: 16:46:50 on Mar 03,2020
# ** Note: (vsim-3812) Design is being optimized...
# //  Questa Sim-64
# //  Version 2019.1 linux_x86_64 Jan  1 2019
# //
# //  Copyright 1991-2019 Mentor Graphics Corporation
# //  All Rights Reserved.
# //
# //  QuestaSim and its associated documentation contain trade
# //  secrets and commercial or financial information that are the property of
# //  Mentor Graphics Corporation and are privileged, confidential,
# //  and exempt from disclosure under the Freedom of Information Act,
# //  5 U.S.C. Section 552. Furthermore, this information
# //  is prohibited from disclosure under the Trade Secrets Act,
# //  18 U.S.C. Section 1905.
# //
# Loading std.standard
# Loading std.textio(body)
# Loading ieee.std_logic_1164(body)
# Loading ieee.numeric_std(body)
# Loading work.sample_module_pack(body)
# Loading work.sample_module(impl)#1
# Loading work.sample_module_1(impl)#1
# Loading .../lib/python3.7/site-packages/cocotb/libs/modelsim/libfli.so
#      -.--ns INFO     cocotb.gpi                         ..ib/fli/FliImpl.cpp:1077 in cocotb_init                     cocotb_init called
# 
#      -.--ns INFO     cocotb.gpi                         ../embed/gpi_embed.c:71   in set_program_name_in_venv        Did not detect Python virtual environment. Using system-wide Python interpreter
#      -.--ns INFO     cocotb.gpi                         ../gpi/GpiCommon.cpp:104  in gpi_print_registered_impl       FLI registered
#      0.00ns INFO     cocotb                                      __init__.py:144  in _initialise_testbench           Running on ModelSim for Questa-64 version 2019.1 2019.01
#      0.00ns INFO     cocotb                                      __init__.py:145  in _initialise_testbench           Python interpreter initialized and cocotb loaded!
#      0.00ns INFO     cocotb                                      __init__.py:163  in _initialise_testbench           Running tests with cocotb v1.4.0.dev0 from .../lib/python3.7/site-packages
#      0.00ns INFO     cocotb                                      __init__.py:180  in _initialise_testbench           Seeding Python random module with 1583275611
#      0.00ns INFO     cocotb.regression                         regression.py:189  in initialise                      Found test test_force_release.test_force_release
#      0.00ns INFO     cocotb.regression                         regression.py:417  in execute                         Running test 1/1: test_force_release
#      0.00ns INFO     ..test_force_release.0x7ffff1543ef0       decorators.py:253  in _advance                        Starting test: "test_force_release"
#                                                                                                                      Description: 
#                                                                                                                          Test force and release on simulation handles
#                                                                                                                          
FATAL: We are calling up again
U/I: Unexpected EOF on RPC channel sock2dc2490

** Fatal: Trouble with Simulation Kernel.
#     10.00ns CRITICAL cocotb.gpi                                FliObjHdl.cpp:328  in set_signal_value                Force or release action not supported for FLI.
make[1]: *** [results.xml] Error 218
make[1]: Leaving directory `.../cocotb/tests/test_cases/test_force_release'
make: *** [sim] Error 2
```

--- a/cocotb/share/include/gpi_logging.h
+++ b/cocotb/share/include/gpi_logging.h
@@ -53,10 +53,7 @@ enum gpi_log_levels {
 #define LOG_INFO(...)      gpi_log("cocotb.gpi", GPIInfo,          __FILE__, __func__, __LINE__, __VA_ARGS__);
 #define LOG_WARN(...)      gpi_log("cocotb.gpi", GPIWarning,       __FILE__, __func__, __LINE__, __VA_ARGS__);
 #define LOG_ERROR(...)     gpi_log("cocotb.gpi", GPIError,         __FILE__, __func__, __LINE__, __VA_ARGS__);
-#define LOG_CRITICAL(...)  do { \
-    gpi_log("cocotb.gpi", GPICritical,      __FILE__, __func__, __LINE__, __VA_ARGS__); \
-    exit(1); \
-} while (0)
+#define LOG_CRITICAL(...)  gpi_log("cocotb.gpi", GPICritical,      __FILE__, __func__, __LINE__, __VA_ARGS__);
 
 void set_log_handler(void *handler);
 void clear_log_handler(void);

--- a/cocotb/share/lib/fli/FliImpl.cpp
+++ b/cocotb/share/lib/fli/FliImpl.cpp
@@ -1016,6 +1016,8 @@ void handle_fli_callback(void *data)
 
     if (!cb_hdl) {
         LOG_CRITICAL("FLI: Callback data corrupted: ABORTING");
+        gpi_embed_end();
+        return;
     }
 
     gpi_cb_state_e old_state = cb_hdl->get_call_state();

--- a/cocotb/share/lib/fli/FliObjHdl.cpp
+++ b/cocotb/share/lib/fli/FliObjHdl.cpp
@@ -83,7 +83,7 @@ int FliObjHdl::initialise(std::string &name, std::string &fq_name)
             m_num_elems = 1;
             break;
         default:
-            LOG_CRITICAL("Invalid object type for FliObjHdl. (%s (%s))", name.c_str(), get_type_str());
+            LOG_ERROR("Invalid object type for FliObjHdl. (%s (%s))", name.c_str(), get_type_str());
             return -1;
     }
 
@@ -228,7 +228,7 @@ long FliEnumObjHdl::get_signal_value_long()
 int FliEnumObjHdl::set_signal_value(const long value, const gpi_set_action_t action)
 {
     if (action != GPI_DEPOSIT) {
-        LOG_CRITICAL("Force or release action not supported for FLI.");
+        LOG_ERROR("Force or release action not supported for FLI.");
         return -1;
     }
 
@@ -269,7 +269,7 @@ int FliLogicObjHdl::initialise(std::string &name, std::string &fq_name)
             }
             break;
         default:
-            LOG_CRITICAL("Object type is not 'logic' for %s (%d)", name.c_str(), m_fli_type);
+            LOG_ERROR("Object type is not 'logic' for %s (%d)", name.c_str(), m_fli_type);
             return -1;
     }
 
@@ -306,7 +306,7 @@ const char* FliLogicObjHdl::get_signal_value_binstr()
             }
             break;
         default:
-            LOG_CRITICAL("Object type is not 'logic' for %s (%d)", m_name.c_str(), m_fli_type);
+            LOG_ERROR("Object type is not 'logic' for %s (%d)", m_name.c_str(), m_fli_type);
             return NULL;
     }
 
@@ -318,7 +318,7 @@ const char* FliLogicObjHdl::get_signal_value_binstr()
 int FliLogicObjHdl::set_signal_value(const long value, const gpi_set_action_t action)
 {
     if (action != GPI_DEPOSIT) {
-        LOG_CRITICAL("Force or release action not supported for FLI.");
+        LOG_ERROR("Force or release action not supported for FLI.");
         return -1;
     }
 
@@ -351,7 +351,7 @@ int FliLogicObjHdl::set_signal_value(const long value, const gpi_set_action_t ac
 int FliLogicObjHdl::set_signal_value_binstr(std::string &value, const gpi_set_action_t action)
 {
     if (action != GPI_DEPOSIT) {
-        LOG_CRITICAL("Force or release action not supported for FLI.");
+        LOG_ERROR("Force or release action not supported for FLI.");
         return -1;
     }
 
@@ -435,7 +435,7 @@ long FliIntObjHdl::get_signal_value_long()
 int FliIntObjHdl::set_signal_value(const long value, const gpi_set_action_t action)
 {
     if (action != GPI_DEPOSIT) {
-        LOG_CRITICAL("Force or release action not supported for FLI.");
+        LOG_ERROR("Force or release action not supported for FLI.");
         return -1;
     }
 
@@ -474,7 +474,7 @@ double FliRealObjHdl::get_signal_value_real()
 int FliRealObjHdl::set_signal_value(const double value, const gpi_set_action_t action)
 {
     if (action != GPI_DEPOSIT) {
-        LOG_CRITICAL("Force or release action not supported for FLI.");
+        LOG_ERROR("Force or release action not supported for FLI.");
         return -1;
     }
 
@@ -522,7 +522,7 @@ const char* FliStringObjHdl::get_signal_value_str()
 int FliStringObjHdl::set_signal_value_str(std::string &value, const gpi_set_action_t action)
 {
     if (action != GPI_DEPOSIT) {
-        LOG_CRITICAL("Force or release action not supported for FLI.");
+        LOG_ERROR("Force or release action not supported for FLI.");
         return -1;
     }
 

--- a/cocotb/share/lib/vhpi/VhpiImpl.cpp
+++ b/cocotb/share/lib/vhpi/VhpiImpl.cpp
@@ -932,8 +932,11 @@ void handle_vhpi_callback(const vhpiCbDataT *cb_data)
 {
     VhpiCbHdl *cb_hdl = (VhpiCbHdl*)cb_data->user_data;
 
-    if (!cb_hdl)
-        LOG_CRITICAL("VHPI: Callback data corrupted");
+    if (!cb_hdl) {
+        LOG_CRITICAL("VHPI: Callback data corrupted: ABORTING");
+        gpi_embed_end();
+        return;
+    }
 
     gpi_cb_state_e old_state = cb_hdl->get_call_state();
 

--- a/cocotb/share/lib/vpi/VpiCbHdl.cpp
+++ b/cocotb/share/lib/vpi/VpiCbHdl.cpp
@@ -164,7 +164,8 @@ int VpiArrayObjHdl::initialise(std::string &name, std::string &fq_name) {
         }
 
         if (rangeHdl == NULL) {
-            LOG_CRITICAL("Unable to get range for indexable object");
+            LOG_ERROR("Unable to get range for indexable object");
+            return -1;
         } else {
             vpi_free_object(iter); // Need to free iterator since exited early
 
@@ -185,7 +186,8 @@ int VpiArrayObjHdl::initialise(std::string &name, std::string &fq_name) {
         check_vpi_error();
         m_range_right = val.value.integer;
     } else {
-        LOG_CRITICAL("Unable to get range for indexable object");
+        LOG_ERROR("Unable to get range for indexable object");
+        return -1;
     }
 
     /* vpiSize will return a size that is incorrect for multi-dimensional arrays so use the range
@@ -260,7 +262,8 @@ int VpiSignalObjHdl::initialise(std::string &name, std::string &fq_name) {
                         check_vpi_error();
                         m_range_right = val.value.integer;
                     } else {
-                        LOG_CRITICAL("Unable to get range for indexable object");
+                        LOG_ERROR("Unable to get range for indexable object");
+                        return -1;
                     }
                 }
                 else {

--- a/cocotb/share/lib/vpi/VpiCbHdl.cpp
+++ b/cocotb/share/lib/vpi/VpiCbHdl.cpp
@@ -96,11 +96,13 @@ int VpiCbHdl::cleanup_callback()
 
     if (m_state == GPI_PRIMED) {
         if (!m_obj_hdl) {
-            LOG_CRITICAL("VPI: passed a NULL pointer : ABORTING");
+            LOG_ERROR("VPI: passed a NULL pointer");
+            return -1;
         }
 
         if (!(vpi_remove_cb(get_handle<vpiHandle>()))) {
-            LOG_CRITICAL("VPI: unable to remove callback : ABORTING");
+            LOG_ERROR("VPI: unable to remove callback");
+            return -1;
         }
 
         check_vpi_error();
@@ -108,7 +110,8 @@ int VpiCbHdl::cleanup_callback()
 #ifndef MODELSIM
         /* This is disabled for now, causes a small leak going to put back in */
         if (!(vpi_free_object(get_handle<vpiHandle>()))) {
-            LOG_CRITICAL("VPI: unable to free handle : ABORTING");
+            LOG_ERROR("VPI: unable to free handle");
+            return -1;
         }
 #endif
     }
@@ -461,7 +464,8 @@ int VpiValueCbHdl::cleanup_callback()
     /* This is a recurring callback so just remove when
      * not wanted */
     if (!(vpi_remove_cb(get_handle<vpiHandle>()))) {
-        LOG_CRITICAL("VPI: unbale to remove callback : ABORTING");
+        LOG_ERROR("VPI: unable to remove callback");
+        return -1;
     }
 
     m_obj_hdl = NULL;

--- a/cocotb/share/lib/vpi/VpiImpl.cpp
+++ b/cocotb/share/lib/vpi/VpiImpl.cpp
@@ -572,6 +572,8 @@ int32_t handle_vpi_callback(p_cb_data cb_data)
 
     if (!cb_hdl) {
         LOG_CRITICAL("VPI: Callback data corrupted: ABORTING");
+        gpi_embed_end();
+        return -1;
     }
 
     gpi_cb_state_e old_state = cb_hdl->get_call_state();

--- a/cocotb/share/lib/vpi/VpiImpl.cpp
+++ b/cocotb/share/lib/vpi/VpiImpl.cpp
@@ -543,8 +543,7 @@ GpiCbHdl *VpiImpl::register_nexttime_callback()
 
 int VpiImpl::deregister_callback(GpiCbHdl *gpi_hdl)
 {
-    gpi_hdl->cleanup_callback();
-    return 0;
+    return gpi_hdl->cleanup_callback();
 }
 
 // If the Python world wants things to shut down then unregister
@@ -587,13 +586,15 @@ int32_t handle_vpi_callback(p_cb_data cb_data)
 
         /* We have re-primed in the handler */
         if (new_state != GPI_PRIMED)
-            if (cb_hdl->cleanup_callback())
+            if (cb_hdl->cleanup_callback()) {
                 delete cb_hdl;
+            }
 
     } else {
         /* Issue #188: This is a work around for a modelsim */
-        if (cb_hdl->cleanup_callback())
+        if (cb_hdl->cleanup_callback()) {
             delete cb_hdl;
+        }
     }
 
     return rv;

